### ISSUE TITLE
Possiblity of bug existence in CrossentropySoftmaxArgmax1HotWithBias

### ISSUE THREAD

**Issue by Amir-Arsalan** [Theano/Theano#3920]
I realized about the issue after trying to answer this [question](http://stackoverflow.com/questions/34567454/theano-how-to-give-training-data-to-a-neural-network) on StackoverFlow.

As you can see in the question page, there are two separate code snippets in my answer. The main difference between the first and the second one is that the former takes the advantage of the built-in cross-entropy function in Theano and the later do not. The question is asking for a working version of the code for a binary classification problem. That is why I wrote the second code snippet, which is only suitable for binary-class problems.

When I had posted the first code snippet for the question, the person who has asked the question kept complaining about my answer and would say he/she gets an error when trying to run the code. I became so curious on what the problem could be and switched from GPU to CPU to run the code. After running, I would get the same errors as the person who has asked the question. I discussed the issue with [Daniel Renshaw] (http://stackoverflow.com/users/127480/daniel-renshaw) and he suggested me to post the issue here. He told me there potentially is a bug in the built-in cross-entropy function. Then I wrote the second piece of code which works fine on both CPU and GPU and kind of confirms the possibility of a bug in CrossentropySoftmaxArgmax1HotWithBias.

Below is my code that does not work on the CPU:

```
import theano
import theano.tensor as T
import numpy


class Layer(object):
    """
    this is a layer in the mlp
    it's not meant to predict the outcome hence it does not compute a loss.
    apply the functions for negative log likelihood = cost on the output of the last layer
    """

    def __init__(self, input, n_in, n_out):
        self.x = input
        self.W = theano.shared(
                value=numpy.zeros(
                        (n_in, n_out),
                        dtype=theano.config.floatX
                ),
                name="W",
                borrow=True
        )
        self.b = theano.shared(
                value=numpy.zeros(n_out,
                                  dtype=theano.config.floatX),
                name="b",
                borrow=True
        )

        self.output = T.nnet.softmax(T.dot(self.x, self.W) + self.b)
        self.params = [self.W, self.b]
        self.input = input


def y_pred(output):
    return T.argmax(output, axis=1)


def negative_log_likelihood(output, y):
    return -T.mean(T.log(output)[T.arange(y.shape[0]), y])


def errors(output, y):
    # check if y has same dimension of y_pred
    if y.ndim != y_pred(output).ndim:
        raise TypeError(
                'y should have the same shape as self.y_pred',
                ('y', y.type, 'y_pred', y_pred(output).type)
        )
    # check if y is of the correct datatype
    if y.dtype.startswith('int'):
        # the T.neq operator returns a vector of 0s and 1s, where 1
        # represents a mistake in prediction
        return T.mean(T.neq(y_pred(output), y))
    else:
        raise NotImplementedError()

data_x = numpy.matrix([[0, 0],
                       [1, 0],
                       [0, 1],
                       [1, 1]])

data_y = numpy.array([0,
                      0,
                      0,
                      1])

train_set_x = theano.shared(numpy.asarray(data_x,
                         dtype=theano.config.floatX),
                         borrow=True)

train_set_y = T.cast(theano.shared(numpy.asarray(data_y,
                         dtype=theano.config.floatX),
                         borrow=True),"int32")

x = T.matrix("x")  # data
y = T.ivector("y")  # labels

classifier = Layer(input=x, n_in=2, n_out=1)

cost = negative_log_likelihood(classifier.output, y)

g_W = T.grad(cost=cost, wrt=classifier.W)
g_b = T.grad(cost=cost, wrt=classifier.b)
index = T.iscalar()

learning_rate = 0.15

updates = (
    (classifier.W, classifier.W - learning_rate * g_W),
    (classifier.b, classifier.b - learning_rate * g_b)
)

train_model = theano.function(
        inputs=[index],
        outputs=cost,
        updates=updates,
        givens={
            x: train_set_x[index:index + 1],
            y: train_set_y[index:index + 1]
        }
)
validate_model = theano.function(
        inputs=[index],
        outputs=errors(classifier.output, y),
        givens={
            x: train_set_x[index:index + 1],
            y: train_set_y[index:index + 1]
        }
)

#train the model
for i in range(train_set_x.shape[0].eval()):
    train_model(i)
```

**Comment by lamblin** [comment]
As mentioned on the updated stackoverflow question, the issue seems to be coming from specifying `n_out=1` when you really have 2 output classes.
I'm closing this ticket, please reopen it if I did not understand what the issue is.

**Comment by Amir-Arsalan** [comment]
@lamblin I cannot re-open the issue since I did not close it. Regarding the issue, the problem is the person who had asked the question was trying to construct a neural network with two outputs. This is still a binary problem but implemented using Softmax function. So suppose one would like to have two output layers instead of one for his/her binary classification problem which has happened here, unintentionally.The problem is not solvable simply by selecting the output classes to 1. If the problem is simply having two classes, why would it run with no issues on the CPU?

**Comment by lamblin** [comment]
If you want to predict independent binary labels, you should use `sigmoid` instead of `softmax`.
A softmax with only one output class will always predict 1, since the output is normalized. You will not be able to train your network from that, even if it appears to run on CPU.
